addons/l10n_in_edi/models/account_edi_format.py

```diff
@@ -0,0 +1,442 @@
+        if self.code == "in_einvoice_1_03":
+            return True
+        return super()._is_enabled_by_default_on_journal(journal)
+
+    def _is_required_for_invoice(self, invoice):
+        self.ensure_one()
+        if self.code == "in_einvoice_1_03" and invoice.is_sale_document():
+            return invoice.l10n_in_gst_treatment in (
+                "regular",
+                "composition",
+                "overseas",
+                "special_economic_zone",
+                "deemed_export",
+            )
+        return super()._is_required_for_invoice(invoice)
+
+    def _needs_web_services(self):
+        self.ensure_one()
+        return self.code == "in_einvoice_1_03" and True or super()._needs_web_services()
+
+    def _get_invoice_edi_content(self, move):
+        if self.code != "in_einvoice_1_03":
+            return super()._get_invoice_edi_content(move)
+        json_dump = json.dumps(self._l10n_in_edi_generate_json(move))
+        return json_dump.encode()
+
+    def _check_move_configuration(self, move):
+        if self.code != "in_einvoice_1_03":
+            return super()._check_move_configuration(move)
+        error_message = []
+        error_message += self._l10n_in_validate_partner(move.partner_id)
+        error_message += self._l10n_in_validate_partner(move.company_id.partner_id, is_company=True)
+        if not move.company_id.l10n_in_edi_username or not move.company_id.sudo().l10n_in_edi_password:
+            error_message.append(_(
+                "A username and password still needs to be set for the E-invoice(IN)."
+                "\nIt needs to be added in the Settings."))
+        if not re.match("^.{1,16}$", move.name):
+            error_message.append(_("Invoice number should not be more than 16 characters"))
+        for line in move.invoice_line_ids.filtered(lambda line: not line.display_type):
+            if line.product_id:
+                if not line.product_id.l10n_in_hsn_code:
+                    error_message.append(_("HSN code is not set in product %s", line.product_id.name))
+                elif not re.match("^[0-9]+$", line.product_id.l10n_in_hsn_code):
+                    error_message.append(_(
+                        "Invalid HSN Code (%s) in product %s", line.product_id.l10n_in_hsn_code, line.product_id.name
+                    ))
+            else:
+                error_message.append(_("product is required to get HSN code"))
+        return error_message
+
+    def _l10n_in_edi_get_iap_buy_credits_message(self, company):
+        base_url = "https://iap-sandbox.odoo.com/iap/1/credit" if not company.l10n_in_edi_production_env else ""
+        url = self.env["iap.account"].get_credits_url(service_name="l10n_in_edi", base_url=base_url)
+        return """<p><b>%s</b></p><p>%s</p>""" % (
+            _("You have insufficient credits to send this document!"),
+            _("Please proceed to buy more credits <a href='%s'>here.</a>", html_escape(url)),
```

Comment:
I'm pretty sure this message is already returned by default by the IAP request. Any chance to have this as a generic message?

Reply:
we handle exceptions in the server and return some static responses so we again handle that here based on that static responses.
Same this done in other EDI so we can put this method in account.edi.formate where we pass service name and test or not.

Reply:
the link has to move inside the markup definition (and you can remove the `html_escape` then)

Reply:
okay